### ISSUE TITLE
rule #196 is hard to understand

### ISSUE THREAD

**Issue by tlossen** [zalando/restful-api-guidelines#302]
> Must: Ensure Event Type schemas conform to Open API’s Schema Object [196]

maybe an explanation of "Open API's Schema Object" (as opposed to Open API itself) would help.

also, i stumbled over "the list of non-available keywords relative to JSON-Schema". to a casual reader, this sounds like `oneOf` is forbidden, while i suppose it simply means that `oneOf` is a reserved keyword.

**Comment by dehora** [comment]
> this sounds like oneOf is forbidden, while i suppose it simply 
> means that oneOf is a reserved keyword.

oneOf isn't available in Open API/swagger 2.0 (it's part of JSON-Schema and that can cause confusion because Nakadi is based on JSON-Schema not Open API, and so you can end up with event types that are not expressible using Open API). oneOf has been added to to v3.0, but the guidelines aren't on that version yet.

**Comment by conorclifford** [comment]
That this rule states that Open API is mandated for Event definitions, and the fact we use Nakadi for event infrastructure, seem to be incompatible - Nakadi does not accept events defined in Open API, but rather JSON Schema (Draft 4 I believe) - there are significant differences (for example, how object polymorphism is modelled very differently between the two) that make the 2 different modelling systems incompatible, and this rule essentially broken IMO.

**Comment by dehora** [comment]
The rule is hard to follow. Here's the intent.

- The JSON Schema used to define the event should be constrained to follow Open API as far as possible because that's the defacto definition format used in APIs.

- That means some things (keywords) should be avoided for the event to fit closer to Open API.

- It also means other things not covered by the rule (eg Open API eg `discriminator`) are impacted.

> this rule essentially broken IMO.

It's reflecting problems with underlying format technology selections, which are incoherent. To narrow it down a bit more, Nakadi's choice of JSON Schema is in this sense, "broken", with respect to the earlier choice to use Open API. Neither are particularly great building blocks when it comes to schema definitions but that's another discussion, and Open API serves well enough as an API description language for social purposes.

What the guidelines would see to like is a world where events were defined as if (or exactly if) they were described in Open API to have uniformity. But, unless Zalando either allow JSON Schema generally for use in APIs definitions, or it has the Nakadi service either, a) enforce the ad-hoc constraints on JSON Schema mentioned in 196, b) accept actual Open API definitions, then there needs to be something to "manage down" the lack of uniformity and divergence between API and event descriptions (Zalando has far more event definitions than APIs today). So until there's a better option (or a better worded rule), this is it unfortunately. Another alternative it to let events and api object go their own way and work with the actual standards that are supported.

**Comment by conorclifford** [comment]
By restricting one technology (JSON Schema) to the subset of another technology (Open API) we end up with a situation where modelling constructs from the former cannot be used, while the equivalent/replacement mechanisms of the latter are not available. 

Within the team I work with, we've run into this with the handling of polymorphic objects - in JSON Schema, this can be done in a mechanism equivalent to that of Open API - even including enumerated fields that acts as a defacto discriminator for precise determinism. However, this still relies on the use of `oneOf` for the "base" object.

Disallowing `oneOf` in favour of Open API's `discriminator` is not an option however, as you rightly point out, Nakadi doesn't provide the capability to validate/process such schemas. This then leaves us in a position of not being able to compliantly model this data.
